### ISSUE TITLE
Computing difference between two documents from the same update leads to non-empty diff

### ISSUE THREAD

**Issue by tmathew0309** [yjs/yjs#346]
**Describe the bug**
 Not sure if I'm doing something wrong but I always seem to get a non-empty update even when calculating differences between two documents that should be the exact same (applied from the exact same update).

```js
const update = new Uint8Array([
4,
3,
254,
234,
225,
208,
13,
0,
40,
0,
235,
182,
241,
188,
3,
0,
2,
105,
100,
1,
119,
36,
55,
56,
49,
56,
100,
99,
57,
99,
45,
50,
50,
50,
54,
45,
52,
49,
97,
55,
45,
97,
57,
49,
101,
45,
57,
53,
51,
57,
53,
51,
102,
97,
101,
57,
56,
98,
4,
0,
235,
182,
241,
188,
3,
3,
1,
97,
132,
254,
234,
225,
208,
13,
1,
1,
98,
52,
130,
219,
255,
228,
5,
0,
132,
159,
176,
193,
155,
3,
8,
1,
32,
132,
130,
219,
255,
228,
5,
0,
1,
104,
132,
130,
219,
255,
228,
5,
1,
1,
101,
132,
130,
219,
255,
228,
5,
2,
1,
114,
132,
130,
219,
255,
228,
5,
3,
1,
101,
132,
130,
219,
255,
228,
5,
4,
1,
32,
132,
130,
219,
255,
228,
5,
5,
1,
105,
132,
130,
219,
255,
228,
5,
6,
1,
115,
132,
130,
219,
255,
228,
5,
7,
1,
32,
132,
130,
219,
255,
228,
5,
8,
1,
97,
132,
130,
219,
255,
228,
5,
9,
1,
32,
132,
130,
219,
255,
228,
5,
10,
1,
98,
132,
130,
219,
255,
228,
5,
11,
1,
117,
132,
130,
219,
255,
228,
5,
12,
1,
110,
132,
130,
219,
255,
228,
5,
13,
1,
110,
132,
130,
219,
255,
228,
5,
14,
1,
118,
132,
130,
219,
255,
228,
5,
15,
1,
104,
132,
130,
219,
255,
228,
5,
16,
1,
32,
132,
130,
219,
255,
228,
5,
17,
1,
99,
132,
130,
219,
255,
228,
5,
18,
1,
104,
132,
130,
219,
255,
228,
5,
19,
1,
32,
132,
130,
219,
255,
228,
5,
20,
1,
109,
132,
130,
219,
255,
228,
5,
21,
1,
111,
132,
130,
219,
255,
228,
5,
22,
1,
114,
132,
130,
219,
255,
228,
5,
23,
1,
101,
132,
130,
219,
255,
228,
5,
24,
1,
32,
132,
130,
219,
255,
228,
5,
25,
1,
116,
132,
130,
219,
255,
228,
5,
26,
1,
101,
132,
130,
219,
255,
228,
5,
27,
1,
120,
132,
130,
219,
255,
228,
5,
28,
1,
116,
132,
130,
219,
255,
228,
5,
29,
1,
32,
132,
130,
219,
255,
228,
5,
30,
1,
116,
132,
130,
219,
255,
228,
5,
31,
1,
104,
132,
130,
219,
255,
228,
5,
32,
1,
97,
132,
130,
219,
255,
228,
5,
33,
1,
116,
132,
130,
219,
255,
228,
5,
34,
1,
32,
132,
130,
219,
255,
228,
5,
35,
1,
73,
132,
130,
219,
255,
228,
5,
36,
1,
105,
132,
130,
219,
255,
228,
5,
37,
1,
39,
132,
130,
219,
255,
228,
5,
38,
1,
118,
132,
130,
219,
255,
228,
5,
39,
1,
101,
132,
130,
219,
255,
228,
5,
40,
1,
39,
132,
130,
219,
255,
228,
5,
41,
1,
118,
132,
130,
219,
255,
228,
5,
42,
1,
101,
132,
130,
219,
255,
228,
5,
43,
1,
32,
132,
130,
219,
255,
228,
5,
44,
1,
119,
132,
130,
219,
255,
228,
5,
45,
1,
114,
132,
130,
219,
255,
228,
5,
46,
1,
105,
132,
130,
219,
255,
228,
5,
47,
1,
116,
132,
130,
219,
255,
228,
5,
48,
1,
116,
132,
130,
219,
255,
228,
5,
49,
1,
101,
132,
130,
219,
255,
228,
5,
50,
1,
110,
5,
235,
182,
241,
188,
3,
0,
7,
1,
7,
99,
111,
110,
116,
101,
110,
116,
1,
39,
0,
235,
182,
241,
188,
3,
0,
8,
99,
104,
105,
108,
100,
114,
101,
110,
0,
7,
0,
235,
182,
241,
188,
3,
1,
1,
39,
0,
235,
182,
241,
188,
3,
2,
4,
116,
101,
120,
116,
2,
40,
0,
235,
182,
241,
188,
3,
0,
4,
116,
121,
112,
101,
1,
119,
9,
112,
97,
114,
97,
103,
114,
97,
112,
104,
9,
159,
176,
193,
155,
3,
0,
132,
254,
234,
225,
208,
13,
2,
1,
99,
132,
159,
176,
193,
155,
3,
0,
1,
100,
132,
159,
176,
193,
155,
3,
1,
1,
101,
132,
159,
176,
193,
155,
3,
2,
1,
102,
132,
159,
176,
193,
155,
3,
3,
1,
103,
132,
159,
176,
193,
155,
3,
4,
1,
104,
132,
159,
176,
193,
155,
3,
5,
1,
105,
132,
159,
176,
193,
155,
3,
6,
1,
106,
132,
159,
176,
193,
155,
3,
7,
1,
107,
1,
130,
219,
255,
228,
5,
2,
14,
4,
37,
4,
]);
const serverDocument = new Doc();
applyUpdate(serverDocument, update);
const clientDocument = new Doc();
applyUpdate(clientDocument, update);
// expected Uint8Array [0,0] here but got non-empty update
const diffsServer = encodeStateAsUpdate(clientDocument, encodeStateVector(serverDocument));
// expected Uint8Array [0,0] here but got non-empty update
const diffsClient = encodeStateAsUpdate(serverDocument, encodeStateVector(clientDocument));
```

**Expected behavior**
Expected both diffsServer and diffsClient to equal Uint8Array [0,0]

**Environment Information**
 - Yjs 13.5.18
 - Chrome Version 95.0.4638.69 (Official Build) (x86_64)

**Comment by dmonad** [comment]
Hi @tmathew0309,

Deletes are not tracked by state vectors, hence clients always exchange all deletes with each other. This is encoded very efficiently and it does not make sense to optimize this further. Not tracking this kind of information (we don't have to assign a lamport timestamp to each delete operation) also allows us to do further compressing of the data. The advantage is that we have a much smaller initial load. The disadvantage is that we always have a small load of deletes even when we are fully synced.

In short, this is the expected behavior expected.

**Comment by tmathew0309** [comment]
Ah I see, thanks a lot for the explanation @dmonad. Will close this then.

**Comment by tmathew0309** [comment]
@dmonad one thing I realized is I was implicitly using the condition that `diff == Uint8Array [0,0]` to indicate that the server document is equal to the client one. Is there a way to efficiently figure out whether two documents are equal without sending the whole document across? Would comparing the state vectors of the server and client documents be sufficient for that?

**Comment by dmonad** [comment]
You could compare snapshots (`Y.snapshot(ydoc)`). They contain the state vector and the delete set. Alternatively, you could create a hash on the snapshot and only compare that if you are only interested in equivalence.

**Comment by tmathew0309** [comment]
Nice that makes sense @dmonad. Is it possible to get the snapshot using only a document as Uint8Array? Something similar to `encodeStateVectorFromUpdate`?

**Comment by tmathew0309** [comment]
Alternatively, would computing a hash on the server and client updates (the result of `encodeStateAsUpdate(doc)`)  and comparing those also work for an equality check?

**Comment by dmonad** [comment]
> Nice that makes sense @dmonad. Is it possible to get the snapshot using only a document as Uint8Array (without loading document in to memory)? Something similar to encodeStateVectorFromUpdate?

Not at the moment. We could add this feature. 

> Alternatively, would computing a hash on the server and client updates (the result of encodeStateAsUpdate(doc)) and comparing those also work for an equality check?

Under very specific conditions this will work. It will definitely fail if you use the alternative update API that only work on the binary format. And it will fail if you use an Undo Manager that keeps changes. I recommend to use the other approach for your sanity. You can also make a hash over the snapshot value to save a few more bytes.
